lib/ZcapClient.js

```diff
@@ -0,0 +1,228 @@
+    allowedActions = (typeof allowedActions === 'string') ?
+      allowedActions = [allowedActions] : allowedActions;
+
+    // default expiration is 5 minutes in the future
+    const defaultExpires = new Date(Date.now() + 5 * 60 * 1000).toISOString();
+
+    if(url) {
+      // generate the root capability
+      delegatedCapability = {
+        '@context': SECURITY_CONTEXT_V2_URL,
+        id: `urn:zcap:${await generateId()}`,
+        parentCapability: 'urn:zcap:' + url,
+        invocationTarget: url,
+        controller: delegate,
+        expires: expires || defaultExpires
+      };
+    } else {
+      delegatedCapability = {
+        // use a provided capability
+        '@context': SECURITY_CONTEXT_V2_URL,
+        id: `urn:zcap:${await generateId()}`,
+        parentCapability: capability.id,
+        invocationTarget: capability.invocationTarget,
+        controller: delegate,
+        expires: expires || defaultExpires
+      };
+    }
+
+    if(allowedActions.length > 0) {
+      delegatedCapability.allowedActions = allowedActions;
+    }
+
+    const signedDelegatedCapability = await jsigs.sign(
+      delegatedCapability, {
+        suite: new Ed25519Signature2018({
+          signer: this.delegationSigner,
+          verificationMethod: this.delegationSigner.id
+        }),
+        purpose: new CapabilityDelegation({
+          capabilityChain: [
+            'urn:zcap:' + url
+          ]
+        })
+      });
+
+    return signedDelegatedCapability;
+  }
+
+  /**
+   * Invokes an Authorization Capability against a given URL.
+   *
+   * @param {object} options - The options to use.
+   * @param {string} options.url - The relative URL to invoke the
+   *   Authorization Capability against.
+   * @param {string} [options.capability] - The capability to invoke at the
+   *   given URL. Default: generate root capability from options.url.
+   * @param {string} [options.method] - The HTTP method to use when accessing
+   *   the resource. Default: 'get'.
+   * @param {string} [options.action] - The capability action that is being
+   *   invoked. Default: 'read'.
+   * @param {object} [options.headers] - The additional headers to sign and
+   *   send along with the HTTP request. Default: {}.
+   * @param {object} options.json - The JSON object, if any, to send with the
+   *   request.
+   *
+   * @returns {Promise<object>} - A promise that resolves to an HTTP response.
+   */
+  async invoke({
+    url,
+    capability,
+    method = 'get',
+    action = 'read',
+    headers = {},
+    json
+  } = {}) {
+    const {baseUrl, agent} = this;
+    const absUrl = `${baseUrl}${url}`;
+
+    // sign the zcap headers
+    const signatureHeaders = await signCapabilityInvocation({
+      url: absUrl,
+      method,
+      headers: {
+        ...headers,
+        date: new Date().toUTCString()
+      },
+      json,
+      invocationSigner: this.invocationSigner,
+      capability: capability || 'urn:zcap:' + absUrl,
```

Comment:
We should probably have a helper function for the `urn:zcap:` prefix stuff.

```suggestion
      capability: capability || `urn:zcap:${absUrl}`,
```